### DISCUSSION TITLE
Update constraints file for 2.3.4

### DISCUSSION THREAD

**Discussion by erdos2n** [apache/airflow#29596] [General]
### What do you see as an issue?

Apache Airflow 2.3.4 constraints file uses

```
apache-airflow-providers-amazon==5.0.0
```

Which has a circular import error. The latest minor version of v5.x.x should be added instead.

```
apache-airflow-providers-amazon==5.1.0
```

### Solving the problem

Update the constraints.txt file

### Anything else

No

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
What do you mean by circular import  @erdos2n ? We almost never update constraints for past releases and I doubt we have a problem here with constraints otherwise we would have thousands of people screaming and our image would not have passed the tests.  Could you explain it please? 

Converting it into discussion until then

**Comment by eladkal** [comment]
I believe you are referring to https://github.com/apache/airflow/pull/25810 which was released in Amazon provider 5.1.0 to fix https://github.com/apache/airflow/issues/26107
This is not good enough reason to update the official constraints and potentially cause damage to so many deployments.
See our policy in the docs https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#fixing-constraint-files-at-release-time

You can (and should if you hit by this issue) to update the provider version locally in your deployment.

**Reply by erdos2n** [reply]
@eladkal, yes that is the fix I'm referring to. Why is the version listed as 5.0.0 even though there is a known issue when using the SecretsBackend from AWS? Is this a common pattern in constraints (leaving broken package versions)? I want to make sure I'm understanding why this is preferred over updating the library version (at least on a minor level).

**Reply by eladkal** [reply]
reason for policy is explained in the docs.
every package/software for every release has bugs. If you update constraints every time new version is released you will not get constancy. Productions systems rely on stable setup might not work as expected. Airflow constrains are generated only once just before release. We are not updating it unless with a very good reason.

You can solve this for your own deployment locally. Many organization generate their own set of constraints according to their needs.

**Reply by potiuk** [reply]
Surely. Let me explain.

Constraints are fixed with the packages that were released and tested together when given version of Airlfow is released. Constraints only purpose is to make sure that you have a stable set of dependencies that are guaranteed to be installable together when you install airlfow together with providers. 

Quotes from: https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

> In order to have a repeatable installation (and only for that reason), we also keep a set of “known-to-be-working” constraint files in the constraints-main, constraints-2-0, constraints-2-1 etc. orphan branches and then we create a tag for each released version e.g. constraints-2.5.1. This way, we keep a tested and working set of dependencies.

Those constraints mean "if you use them -then you will be able to repeatably install Airlfow with the very set of dependencies that we determined as working when we released given airflow version.

The same page explains you also that you are completely free to upgrade providers and other dependencies at any time for whatever reason - for example when new provider has been released. Constraints work in the way that they are only temporary applied at the moment `pip install` command is run. You have all the scenarios described including 

> Installing/upgrading/downgrading providers separately from Airflow core
>
> In order to add new features, implement bug-fixes or simply maintain backwards compatibility, you might need to install,
> upgrade or downgrade any of the providers you need. We release providers independently from the core of Airflow, so often
> new versions of providers are released before Airflow is, also if you do not want yet to upgrade Airflow to the latest version,
> you might want to install newly released providers separately. For installing the providers you should not use any constraint
> files (the constraints are for installing Airflow with providers, not to install providers separately).

Answering your question:

> Is this a common pattern in constraints (leaving broken package versions)?

Absolutely and ultimately yes. I think you do not realize the fact of life that no matter what you do, you always have broken packages in your installation. Not only for airflow - this is a fact of life for pretty much any application there.

In case of airflow - we have 650+ dependencies. When you look how often they change we have on average 3-5 updated packages (fixes) a day. This means that at the day we release airflow, 3 -5 of those packages are marked as broken, because there is a new release for them which fixes some bugs. So effectively yes - every set of constraints we have contains broken packages  - they are broken because either there is already a fixed version released for such packages, or we simply do not realise that there is some bug in there.

And we are not going to update the constraints 3 times a day - mostly because "constraints" purpose is not to provide "bug free version of dependencies" (not sure where you took the impression it is like that), but to provide our users a way to repeatably install airfllow - with the exact set of dependencies that were valid at the moment it has been released.

If you want to keep updated with those, bugfixes you have those options:

* upgrade alll the dependencies that get released with fixes manually (including providers) - if you want to be at the "bleeding edge" of providers 
* upgrade to latest released version of Airflow (it will contain constraints with provider versions released at the moment when the newer version was released) - if you want to be at the top of latest Airflow (highly recomend - this will also allow you to apply all the latest security patches also released in latest airflow versions - we only release them in the latest airlfow released - so 2.4.3 wil not get security fixes released in 2.5.*).
